internal/rbd/nodeserver.go

```diff
@@ -1512,32 +1522,99 @@ func (ns *NodeServer) NodeGetVolumeStats(
 	if stat.Mode().IsDir() {
 		return csicommon.FilesystemNodeGetVolumeStats(ctx, ns.Mounter, targetPath, true)
 	} else if (stat.Mode() & os.ModeDevice) == os.ModeDevice {
-		return blockNodeGetVolumeStats(ctx, targetPath)
+		return ns.blockNodeGetVolumeStats(ctx, targetPath, volumeId)
 	}
 
 	return nil, fmt.Errorf("targetpath %q is not a block device", targetPath)
 }
 
 // blockNodeGetVolumeStats gets the metrics for a `volumeMode: Block` type of
-// volume. At the moment, only the size of the block-device can be returned, as
-// there are no secrets in the NodeGetVolumeStats request that enables us to
-// connect to the Ceph cluster.
+// volume. At the moment,
+// - if the volume is dynamically provisioned and secretRef is present in the
+// configmap, both size and used metrics are returned by performing a
+// DiffIterate on the rbd image.
+// - else only the size of the block-device can be returned, as there are no
+// secrets in the NodeGetVolumeStats request that enables us to connect to the
+// Ceph cluster.
 //
 // TODO: https://github.com/container-storage-interface/spec/issues/371#issuecomment-756834471
-func blockNodeGetVolumeStats(ctx context.Context, targetPath string) (*csi.NodeGetVolumeStatsResponse, error) {
-	mp := volume.NewMetricsBlock(targetPath)
-	m, err := mp.GetMetrics()
+func (ns *NodeServer) blockNodeGetVolumeStats(ctx context.Context,
+	targetPath,
+	volumeId string) (*csi.NodeGetVolumeStatsResponse, error) {
+	var (
+		err error
+		vi  util.CSIIdentifier
+	)
+	isDiffPossible := true
+
+	err = vi.DecomposeCSIID(volumeId)
 	if err != nil {
-		err = fmt.Errorf("failed to get metrics: %w", err)
-		log.ErrorLog(ctx, err.Error())
+		// treat it as a static volume if decomposition fails.
+		// Static volumes do not support diff iterate to measure usage.
+		isDiffPossible = false
+	}
 
-		return nil, status.Error(codes.Internal, err.Error())
+	secretName, secretNamespace, err := util.GetControllerPublishSecretRef(volumeId, util.RBDType)
+	if err != nil {
+		// diff iterate to measure usage is not possible without secrets.
+		isDiffPossible = false
```

Comment:
Let's also log it? Would help in debugging.

Reply:
> Let's also log it? Would help in debugging.

NodeGetVolumeStats is a very frequest call like every <~2 mins,
Logging this would easily fill up the entire log.

At this point, I don't think lot of users have added publishsecret to the configmap too.

Reply:
While I agree with that, we would never know why `isDiffPossible` is not possible (no pun intended) xD